### ISSUE TITLE
can a plugin hook to event for the (start/initial letter) of a sequence?

### ISSUE THREAD

**Issue by wis** [ranger/ranger#1533]
I'm trying to write a ranger plugin script that runs the shell command `_beep 200 200`, 
here's the shell function that beeps using the speakers:
```
_beep() {
  ( \speaker-test --frequency "$1" --test sine )&
  pid=$!
  \sleep 0."${2}"s
  \kill -9 $pid
}
```
also can you hook to window focus event? and can you check then if a sequence is started and waiting for a key press?
this will be very useful to stop me from running things accidentally.

**Comment by toonn** [comment]
I doubt terminals pass on window manager events through ncurses. Tbh I don't think anyone knows much about plugins rn, they're not super popular because ranger's so easy to customize in other ways.

You can run something when `a` is pressed without interrupting command composition for, e.g., `ab` or `abc`, by using the special `<bg>` "background" key. So if you want a beep when `a` is pressed, but still allow for longer maps starting with `a`; you'd `map a<bg> shell _beep 200 200` and then `map ab some_other_command`.

**Comment by wis** [comment]
Thanks a lot, I could get this to work:
```
map d<bg> shell ( speaker-test --frequency "1500" --test sine )& pid=$!; sleep "0.075"; kill -9 $pid
map a<bg> shell ( speaker-test --frequency "1000" --test sine )& pid=$!; sleep "0.075"; kill -9 $pid
map g<bg> shell ( speaker-test --frequency "1000" --test sine )& pid=$!; sleep "0.075"; kill -9 $pid
map h<bg> shell ( speaker-test --frequency "1000" --test sine )& pid=$!; sleep "0.075"; kill -9 $pid
map m<bg> shell ( speaker-test --frequency "1000" --test sine )& pid=$!; sleep "0.075"; kill -9 $pid
map Y<bg> shell ( speaker-test --frequency "1000" --test sine )& pid=$!; sleep "0.075"; kill -9 $pid
map y<bg> shell ( speaker-test --frequency "1250" --test sine )& pid=$!; sleep "0.075"; kill -9 $pid
map j<bg> shell ( speaker-test --frequency "1000" --test sine )& pid=$!; sleep "0.075"; kill -9 $pid
map c<bg> shell ( speaker-test --frequency "1000" --test sine )& pid=$!; sleep "0.075"; kill -9 $pid
map b<bg> shell ( speaker-test --frequency "1000" --test sine )& pid=$!; sleep "0.075"; kill -9 $pid
map p<bg> shell ( speaker-test --frequency "1125" --test sine )& pid=$!; sleep "0.075"; kill -9 $pid
map s<bg> shell ( speaker-test --frequency "1250" --test sine )& pid=$!; sleep "0.075"; kill -9 $pid
map t<bg> shell ( speaker-test --frequency "1000" --test sine )& pid=$!; sleep "0.075"; kill -9 $pid
map T<bg> shell ( speaker-test --frequency "1000" --test sine )& pid=$!; sleep "0.075"; kill -9 $pid
map r<bg> shell ( speaker-test --frequency "1500" --test sine )& pid=$!; sleep "0.075"; kill -9 $pid
```
speaker-test runs in the foreground and blocks the following keys until exiting, I found that 75 milliseconds is long enough to hear and short enough not to block the following keys in normal use, the more dangerous commands have higher sine wave frequencies. 
`map <any><bg>` didn't work, and the bash function `_beep` wasn't found even though it's in my .bashrc and i'm sure it was loaded (I ran it from the terminal opened from range)
`map r<bg>`, `map R<bg>` and `map A<bg>` don't work too, Ranger waits for the `<bg>` key strokes instead, will it require monkey-patching/wrapping these (renaming) commands from commands.py, how can it be done?

**Comment by toonn** [comment]
Oh, `:shell` doesn't run an interactive shell so aliases and functions in your bashrc aren't loaded. Easiest way around that is to make it a script rather than a function.

It doesn't work for the single key mappings because those are shorter and therefore match first. You'd have to change those maps using `chain`, i.e. `:map <space> chain tag_toggle; shell _beep`.

**Comment by wis** [comment]
thanks! @toonn, now the only issue is that vim style commands e.g 10<space> to select 10 down don't work/repeat, only 1 is selected.

**Comment by toonn** [comment]
Do you mean `10j`? That'd move down 10 lines. `10v` would mark 10 entries.

**Comment by wis** [comment]
the latter, I meant mark by "select", and I have `v` remapped to `<space>`

```
map <Space> chain shell sbeep "1250" "0.070"; mark_files toggle=True
map va       mark_files all=True toggle=True
map uv      mark_files all=True val=False
map vs       toggle_visual_mode
map uV      toggle_visual_mode reverse=True
```

**Comment by toonn** [comment]
Oic, yeah, the positional argument will be given to chain, which probably ignores it. Don't think that can really be solved sadly. Would be pretty arbitrary to assign it to the last chain argument. But maybe?

**Comment by Vifon** [comment]
At this point I'd recommend writing a simple new command in `commands.py` that calls these two with appropriate arguments. `:chain` is indeed prone to such problems.

**Comment by toonn** [comment]
Good idea. Maybe read the implementation of `chain` and implement a similar wrapper that takes only 1 argument, passes the modifier (`10`) along to that argument *and* beeps. Hmm, would need a second argument for the frequency and maybe duration but the principle still holds.
